cloudsync/smartsync.py

```diff
@@ -0,0 +1,384 @@
+            return None
+        found_ents = set()
+        for ent in ents:
+            found = self._smart_unsync_ent(ent)
+            if found:
+                found_ents.add(found)
+        return found_ents
+
+    def _smart_sync_ent(self, ent: SyncEntry) -> bool:
+        """Request to sync down a file from the cloud, and mark the entry to maintain synchronization."""
+        # if the local file is missing,
+        #   clear the local side of the ent and
+        #   mark the remote side changed and clear the sync_path/sync_hash so that it will look new and sync down
+        with self._mutex:
+            for parent_conflict in self.smgr.get_parent_conflicts(ent, REMOTE):  # ALWAYS remote
+                self._sync_one_entry(parent_conflict)
+            return self._sync_one_entry(ent)
+
+    def smart_sync_oid(self, remote_oid):
+        ent: SyncEntry = self.state.smart_sync_oid(remote_oid)
+        if not ent:
+            raise ex.CloudFileNotFoundError(remote_oid)
+        self._smart_sync_ent(ent)
+        return ent[LOCAL].path
+
+    def smart_sync_path(self, path, side):
+        remote_path = self._ensure_path_remote(path, side)
+        try:
+            ents = self.state.smart_sync_path(remote_path)
+        except ex.CloudException as e:
+            self.nmgr.notify_from_exception(SourceEnum.SYNC, e, remote_path)
+            raise
+        for ent in ents:
+            self._smart_sync_ent(ent)
+
+    def smart_listdir_path(self, local_path):
+        """
+        Returns the listdir calculated by the local listdir mixed with the remote sync state, when a file does
+        not exist locally
+        """
+        # assumes that local is efficient and remote is inefficient
+        # read from cache as much as possible remotely, read from disk as much as possible locally
+        # calls local listdir, local listdir overrides cached info, and also update the state with this info too
+        #   remote_oid and is_synced
+        local, remote = self.providers
+        remote_path = self.translate(REMOTE, local_path)
+        local_ents = dict()
+        remote_ents = dict()
+        for dirent in local.listdir_path(local_path):
+            local_ents[dirent.name] = dirent
+        for ent in self.state.smart_listdir_path(remote_path):
+            remote_ents[remote.basename(ent[REMOTE].path)] = ent
+        names = set(local_ents.keys()).union(remote_ents.keys())
+        for name in names:
+            rent = remote_ents.get(name)
+            lent = local_ents.get(name)
+            yield_val = self._ent_to_smartinfo(rent, lent, local.join(local_path, name))
+            yield yield_val
+
+    def _ensure_path_remote(self, path, side) -> str:
+        if side == LOCAL:
+            path = self.translate(REMOTE, path)  # TODO translate the path to remote
+        return path
+
+    def smart_info_path(self, local_path) -> Optional[SmartInfo]:
+        remote_path = self.translate(REMOTE, local_path)
```

Comment:
Note that this may be `None` in scenarios where a downstream user of the library filters some files out. This will result in an exception in `_ent_to_smartinfo` when doing `name = self.providers[REMOTE].basename(ent[REMOTE].path)`.

Not sure whether it's better to just return early here or handle `ent[REMOTE].path == None` in `_ent_to_smartinfo`.

Reply:
What is the expected behavior if the path is filtered out? Also, does listdir need to run all the names through the filter?

Reply:
A: filter out remote-only files that don't translate, don't filter any files that exist locally

Reply:
done